### PR TITLE
Test certificates success and error cards (EXPOSUREAPP-7439, -7454, -7452)

### PR DESCRIPTION
**Pull request by AlexanderAlferov**
Test certificates cards:

Screenshot 2021-06-03 at 18 55 38

⚠️ Data should be adjusted/replaced during merging with business logic ⚠️

### CODE REVIEW

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/greencertificate/ui/certificates/CertificatesAdapter.kt [hunk 1]
```diff
@@ -37,6 +39,8 @@ class CertificatesAdapter :
                 TypedVHCreatorMod({ data[it] is CreateVaccinationCard.Item }) { CreateVaccinationCard(it) },
                 TypedVHCreatorMod({ data[it] is HeaderInfoVaccinationCard.Item }) { HeaderInfoVaccinationCard(it) },
                 TypedVHCreatorMod({ data[it] is BottomInfoVaccinationCard.Item }) { BottomInfoVaccinationCard(it) },
+                TypedVHCreatorMod({ data[it] is VaccinationTestSuccessCard.Item }) { VaccinationTestSuccessCard(it) },
+                TypedVHCreatorMod({ data[it] is VaccinationTestErrorCard.Item }) { VaccinationTestErrorCard(it) },
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/greencertificate/ui/certificates/CertificatesAdapter.kt:43]: Maybe we should rename the cards here , those cards are not for Vaccination right? 
`CovidTestCertificateCard`,  `CovidTestCertificateErrorCard(it)` ? or whatever you think it is better ?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/greencertificate/ui/certificates/CertificatesViewModel.kt [hunk 2]
```diff
@@ -70,6 +73,22 @@ class CertificatesViewModel @AssistedInject constructor(
                     )
                 )
                 add(BottomInfoVaccinationCard.Item)
+
+                // TODO: replace item params with correct data
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/greencertificate/ui/certificates/CertificatesViewModel.kt:77]: Repository is merged , are you doing the wiring in another PR ?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/greencertificate/ui/certificates/CertificatesViewModel.kt [hunk 3]
```diff
@@ -70,6 +73,22 @@ class CertificatesViewModel @AssistedInject constructor(
...
+                    VaccinationTestSuccessCard.Item(
+                        testDate = Instant.now(),
+                        testPerson = "Andrea Schneider"
+                    )
+                )
+                add(
+                    VaccinationTestErrorCard.Item(
+                        testDate = Instant.now(),
+                        onClickAction = {
+                            // TODO: retry failed test certificate function here
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/greencertificate/ui/certificates/CertificatesViewModel.kt:88]: I think you can inject the repo and try to write thing up

File: Corona-Warn-App/src/main/res/layout/vaccination_test_error_card.xml [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        android:layout_marginStart="24dp"
+        android:layout_marginBottom="17dp"
+        android:layout_marginTop="21dp"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:text="@string/test_certificate_error_retry_button"
+        android:textColor="@color/colorAccent"
+        android:textSize="13sp"
+        app:cornerRadius="4dp"
+        app:rippleColor="#33AAAAAA"
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/res/layout/vaccination_test_error_card.xml:80]: please move the color into resources ?

File: Corona-Warn-App/src/main/res/layout/vaccination_test_error_card.xml [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        app:strokeColor="@color/colorAccent"
+        app:strokeWidth="1dip"
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/res/layout/vaccination_test_error_card.xml:82]: `1dp` ?

File: Corona-Warn-App/src/main/res/layout/vaccination_test_success_card.xml [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    <ImageView
+        android:id="@+id/prc_icon"
+        style="@style/icon"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginBottom="24dp"
+        android:layout_marginStart="23dp"
+        android:focusable="false"
+        android:importantForAccessibility="no"
+        android:src="@drawable/ic_risk_details_pcr"
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/res/layout/vaccination_test_success_card.xml:78]: `srcCompat`?

**Review by mtwalli — COMMENTED** [review]
Regarding the data wiring, I think you can do this already here , or you plan to do this in another PR ?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/greencertificate/ui/certificates/CertificatesAdapter.kt (hunk 1 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/greencertificate/ui/certificates/CertificatesAdapter.kt]: renamed

**Review by mtwalli — COMMENTED** [review]
Lgtm overall 👍🏽

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/greencertificate/ui/certificates/CertificatesViewModel.kt [hunk 7]
```diff
@@ -16,62 +20,106 @@ import de.rki.coronawarnapp.vaccination.ui.cards.CreateVaccinationCard
...
-    private val vaccinationSettings: VaccinationSettings
+    private val vaccinationSettings: VaccinationSettings,
+    private val testCertificateRepository: TestCertificateRepository
 ) : CWAViewModel() {
 
     val events = SingleLiveEvent<CertificatesFragmentEvents>()
 
-    // TODO: cards should be adjusted in the following PR
+    private fun refreshTestCertificate(identifier: TestCertificateIdentifier) {
+        viewModelScope.launch {
```
[the remark this answers is not in the document]
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/greencertificate/ui/certificates/CertificatesViewModel.kt]: ok

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/greencertificate/ui/certificates/CertificatesViewModel.kt [hunk 8]
```diff
@@ -16,62 +20,106 @@ import de.rki.coronawarnapp.vaccination.ui.cards.CreateVaccinationCard
...
+                    }
+                )
+            )
+        }
+    }
 
+    private fun MutableList<CertificatesItem>.addTestCertificateCards(certificates: Set<TestCertificateContainer>) {
+
+        certificates.forEach { certificate ->
+            if (certificate.certificateId == null) {
```
[the remark this answers is not in the document]
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/greencertificate/ui/certificates/CertificatesViewModel.kt]: updated

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/greencertificate/ui/certificates/CertificatesViewModel.kt [hunk 9]
```diff
@@ -16,62 +20,106 @@ import de.rki.coronawarnapp.vaccination.ui.cards.CreateVaccinationCard
...
                 add(
-                    CreateVaccinationCard.Item(
+                    CovidTestCertificateErrorCard.Item(
+                        testDate = Instant.now(),
```
[the remark this answers is not in the document]
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/greencertificate/ui/certificates/CertificatesViewModel.kt]: updated

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/greencertificate/ui/certificates/CertificatesViewModel.kt [hunk 10]
```diff
@@ -16,62 +20,106 @@ import de.rki.coronawarnapp.vaccination.ui.cards.CreateVaccinationCard
...
+                        certificate.toTestCertificate(null)?.firstName + " " +
+                            certificate.toTestCertificate(null)?.lastName
+                    )
+                )
             }
-        }.asLiveData()
+        }
+
+        if (certificates.isEmpty()) {
+            add(BottomInfoVaccinationCard.Item)
```
[the remark this answers is not in the document]
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/greencertificate/ui/certificates/CertificatesViewModel.kt]: updated

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/ui/cards/CovidTestCertificateErrorCard.kt [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+            R.string.test_certificate_time,
+            dateTime.toString(dateFormat),
+            dateTime.toString(timeFormat),
+        )
+
+        retryButton.setOnClickListener { item.onClickAction(item) }
+    }
+
+    data class Item(
+        // TODO: replace with correct data
```
[the remark this answers is not in the document]
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/ui/cards/CovidTestCertificateErrorCard.kt]: no

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/ui/cards/VaccinationTestItem.kt [hunk 12]
```diff
@@ -0,0 +1,12 @@
+package de.rki.coronawarnapp.vaccination.ui.cards
+
+import de.rki.coronawarnapp.greencertificate.ui.certificates.items.CertificatesItem
+import org.joda.time.Instant
+
+interface VaccinationTestItem : CertificatesItem {
+    // TODO should be replaced with something like VacccinatedTestPerson
```
[the remark this answers is not in the document]
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/ui/cards/VaccinationTestItem.kt]: no

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/ui/cards/CovidTestCertificateCard.kt [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+            R.string.test_certificate_time,
+            dateTime.toString(dateFormat),
+            dateTime.toString(timeFormat),
+        )
+
+        personName.text = item.testPerson
+    }
+
+    data class Item(
+        // TODO: replace with correct data
```
[the remark this answers is not in the document]
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/ui/cards/CovidTestCertificateCard.kt]: no

**Review by NataliaLemmerth — DISMISSED** [review]
UA approved

**Review by mtwalli — APPROVED** [review]
Lgtm

**Review by axelherbstreith — APPROVED** [review]
lgtm

**Review by jurajkusnier — APPROVED** [review]

**Review by NataliaLemmerth — COMMENTED** [review]
UA approved

**Pull request merged**